### ISSUE TITLE
[Firefox] Can't modify page properties on sites which use CSP

### ISSUE THREAD

**Issue by chocolateboy** [violentmonkey/violentmonkey#1001]
Re-opening this as a new issue as the last one got derailed...

This is my abiding issue with Violentmonkey (which I :heart: - thank you!), but I can't see an open issue for it. There are related closed issues, and [this issue](https://github.com/violentmonkey/violentmonkey/issues/604) may point the way to a fix, but I thought it'd be better to track the bug explicitly rather than inferring it from the [documentation](https://violentmonkey.github.io/posts/inject-into-context/#mode-content)<sup>[1]</sup> and scattered comments.

### What is the problem?

It's not possible to modify/mutate direct or nested properties of a page's window object with the following combination:

- Violentmonkey for Firefox
- sites which use [CSP](https://bugzilla.mozilla.org/show_bug.cgi?id=1267027) (e.g. GitHub, Google, Twitter)

Sites which use CSP don't run in Violentmonkey for Firefox unless [`@inject-into content`](https://violentmonkey.github.io/posts/inject-into-context/#mode-content) is enabled, but `@inject-into content` is not compatible with [`unsafeWindow`](https://wiki.greasespot.net/index.php?title=UnsafeWindow&oldid=7050) (or [`@grant none`](https://wiki.greasespot.net/index.php?title=@grant&oldid=7438)), which is needed to modify page objects.

There is a [workaround](https://github.com/violentmonkey/violentmonkey/issues/997#issuecomment-637700732) for this on Firefox, but its use is not always obvious, and it requires browser/engine-specific code, which userscripts are meant to eliminate.

### How to reproduce it?

```javascript
// ==UserScript==
// @name          Hook XHR#open
// @version       0.0.1
// @include       https://twitter.com/*
// @include       https://github.com/*
// @include       https://*.google.tld/*
// @inject-into   content
// ==/UserScript==

const xhrProto = unsafeWindow.XMLHttpRequest.prototype

function hookXHROpen (oldOpen) {
    return function open () {
        console.warn('inside XHR#open')
        return oldOpen.apply(this, arguments)
    }
}

xhrProto.open = hookXHROpen(xhrProto.open)
```

### What is the expected result?

`XHR#open` should be hooked and the message should be logged on those sites.

### What is the actual result?

`XHR#open` isn't hooked and the message isn't logged.

### Compatibility

Userscript engines this works in:

- Tampermonkey (tested on Firefox)

Userscript engines this doesn't work in:

- [Greasemonkey 4](https://github.com/greasemonkey/greasemonkey/issues/3015)<sup>[2]</sup>
- FireMonkey<sup>[3]</sup>

### Related issues

- https://github.com/violentmonkey/violentmonkey/issues/107
- https://github.com/violentmonkey/violentmonkey/issues/172
- https://github.com/violentmonkey/violentmonkey/issues/173
- https://github.com/violentmonkey/violentmonkey/issues/408
- https://github.com/violentmonkey/violentmonkey/issues/487
- https://github.com/violentmonkey/violentmonkey/issues/604

### Environment

- Browser: Firefox v76.0.1
- Violentmonkey: v2.12.7
- OS: Linux (Arch)

### Footnotes

1. "Scripts requiring access to JavaScript objects in the web page will not work in [`@inject-into content`] mode."
2. "GM4 does not yet support `@grant none`."
3. `@grant none` isn't supported. [`unsafeWindow` is](https://github.com/erosman/support/issues/80) but I couldn't get the `XHR#open` hook to work.

**Comment by tophf** [comment]
The underlying reason is a bug or an architectural limitation in Firefox, see https://bugzil.la/1267027 ("Page CSP should not apply to content inserted by content scripts").

1. We already tried to set unsafeWindow to `unsafeWindow.wrappedJSObject` but if a script exposes a function directly into unsafeWindow it'll break the page. There are scripts like that, including the popular ones - not that it matters because the only solution is for the authors to use `exportFunction` properly, but that is an unnecessary complication in 99.9999% of cases if we include all sites and all browsers like Chrome where this is not a problem. Realistically, we can't enforce that and unlike Greasemonkey4 we can't "move forward" either as Violentmonkey's purpose is to support the classic scripts.

2. Using browser.userScripts API won't help because this API runs the scripts in what essentially is a `content` mode so the above paragraph applies.

3. **So, the only available workaround for extensions is to remove the CSP header entirely like Tampermonkey does**. Turns out, Firefox doesn't allow graceful relaxation of CSP, it can't add a `nonce` or `unsafe-inline`, it can only make CSP stricter or remove it entirely. Yeah, ridiculous, but probably caused by some architectural limitation as well.

Implementation-wise, removing the CSP header is simple and since Tampermonkey is apparently allowed to do it by the AMO reviewers, we can add such an option as well.

**Comment by chocolateboy** [comment]
> On the other hand, the same may be achieved by any user by setting security.csp.enable to false in about:config

Does the header solution disable CSP for all requests to an affected site? Or is it somehow scoped/sandboxed to a particular userscript?

If the former, it sounds like an insecure and dangerous workaround rather than a solution.

Ditto, of course, the user disabling it globally.

**Comment by tophf** [comment]
CSP header is present only on the main document's response. It's not scoped to a script. It's global to the page environment. It defines what the page can or cannot do. Removing the header on all URLs is the same as disabling CSP in about:config. The only advantage of doing it in Violentmonkey is that we can remove the header only if the site has active userscripts. Admittedly, this advantage is worth adding the workaround, but on the other hand this workaround is so lame that I'm reluctant to add it.

**Comment by spodermenpls** [comment]
@tophf Thank you very much for being so kind and lending GoodTwitter2's author a hand with this issue, that was solved faster than I've imagined!

On a side note: Since I've been back using VM for this very userscript after 2.5 years of no need for it, I've been surprised to see that nobody has even tried maintaining the German locale, I've Iron Manned it in one sitting and updated/corrected about half of the entire thing. with Transifex not being too workflow-friendly compared to Crowdin, I might add.

**Comment by sn260591** [comment]
It is also worth noting that the hack with removing the CSP can cause a conflict with other addons that use CSP to block content (like ublock origin).

**Comment by mzzsfy** [comment]
I encountered similar problems, but I used Chrome and Edge
https://github.com/violentmonkey/violentmonkey/issues/172
The difference is that I cannot execute `new function ()`, and an error message
`EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src github.githubassets.com".`

This is the script I use
`https://greasyfork.org/zh-CN/scripts/419215-%E8%87%AA%E5%8A%A8%E6%97%A0%E7%BC%9D%E7%BF%BB%E9%A1%B5/feedback`

**Comment by tophf** [comment]
@mzzsfy, this issue is only about Firefox because in Chromium-based browsers the userscript authors can use GM_createElement to create DOM scripts.

**Comment by mzzsfy** [comment]
ok, i will open a new issue to ask this question

**Comment by tophf** [comment]
@mzzsfy, there's no need. You can suggest the author to use `GM_addElement` or you can install a separate extension to disable CSP of the site or you can switch to Tampermonkey, which has a built-in option to disable site's CSP.

**Comment by mzzsfy** [comment]
Thank you. I will try to use 'GM_addElement' and contact the author

**Comment by chocolateboy** [comment]
Not sure if this is relevant, but just in case:

> The [following change](https://github.com/uBlockOrigin/uBlock-issues/issues/235#issuecomment-1344313926) makes uBlock immune to the page's CSP

**Comment by tophf** [comment]
Definitely looks like a bug in Firefox. Not sure we should exploit it, but on the other hand I consider the original problem a bug in Firefox, so canceling it with another one is a kind of poetical justice.

**Comment by tophf** [comment]
Ah, I doubt though we can use it because it's asynchronous due to `src` and not `textContent` which we use now to pass the command channel securely so that the site doesn't gain full control of it.

**Comment by tophf** [comment]
We can add a global option "Bypass site's CSP for scripts with `page` injection mode", enabled by default.

The injection mode can be set in the script code via `@inject-into page` by its author or in script's settings page by a user or via the global injection mode option in Violentmonkey's advanced settings (this one will effectively remove CSP for all scripts).

@gera2ld?

**Comment by gera2ld** [comment]
Does that also mean the injection mode `content` can actually be removed? It looks great to me.

p.s. IIRC we have done this trick to inject scripts into Firefox in an early version and successfully caught the reviewer's attention, leaving me with the impression that such practices are not welcome in Firefox, especially when injecting code from outside this extension.

**Comment by erosman** [comment]
> p.s. IIRC we have done this trick to inject scripts into Firefox in an early version and successfully caught the reviewer's attention, leaving me with the impression that such practices are not welcome in Firefox, especially when injecting code from outside this extension.

For reference:

> [Add-on Policies](https://extensionworkshop.com/documentation/publish/add-on-policies/): Development Practices
> 
> - Add-ons must not relax web page security headers, such as the Content Security Policy.

**Comment by tophf** [comment]
We should keep `content` mode because the option may be disabled or the script may want to run in `content` for reliability on a hostile site. When MV3 implements userscripts API we'll only allow `page` and `userscript` (or another name) worlds and `content` will be equal to `userscript`.

Tampermonkey removes the CSP header and is allowed in AMO, so I guess it's okay as long as it's optional and user-controlled.

**Comment by kafene** [comment]
I noticed one of my scripts was breaking again on github with the message "Could not inject some scripts", even though I have removed all CSP headers via [an extension](https://addons.mozilla.org/en-US/firefox/addon/security-csp-enable-false/), maybe Firefox has started blocking this ability and doesn't respect its modifications, or loads it after ViolentMonkey.

I tried TamperMonkey and that "just works" without modification. But I am quite used to ViolentMonkey and prefer it, so I think **ViolentMonkey should include the same option to modify CSP headers itself**.

However, I did manage to fix my script to work, by changing a few things, maybe this will help anyone else who is looking:

```js
// This was previously "@inject-into page"
// @inject-into  content

// I did not change this, but include it so you know what "win" is below.
const win = (typeof unsafeWindow === 'object' && unsafeWindow != null) ? unsafeWindow : window;

// I had to comment most of this out, or got an error "can't access dead object"
// In the content context, sites should not be able to override URL() anyway.
/* hack for sites that override URL() */
const _URL = win.URL; /*(() => {
    let urlDefinition;
    try {
        const iframe = doc.createElement('iframe');
        doc.documentElement.appendChild(iframe);
        urlDefinition = iframe.contentWindow.URL;
        iframe.remove();
    } catch (e) {}
    return (urlDefinition || win.URL);
})();*/

// I had to change this to use exportFunction, because I can not use deinfeProperty on window/unsafeWindow
/*Object.defineProperty(win, 'myExportedFunction', {value: myExportedFunction});*/
exportFunction(function () {
    let result = myExportedFunction();
    /* result will look like:
    {
      "url": "https://github.com/violentmonkey/violentmonkey-oex",
      "title": "violentmonkey/violentmonkey-oex - GitHub",
      "description": "violentmonkey/violentmonkey-oex - Violentmonkey, userscripts support for Opera (Presto).",
      "tags": ["github", "gh-archived"]
    }
    */
    return cloneInto(result, window);
}, window, {defineAs: 'myExportedFunction'});

```

The purpose of my script is to export this function that's used by a bookmarklet. It's far too large to just paste into a bookmarklet, which is why I'm using the userscript to inject a function, which is in turn called by the bookmarklet.

I also found that `cloneInto` does not work with some objects. I originally had the `url` property as an object created by `new _URL()`, which caused another error "Encountered unsupported value type writing stack-scoped structured clone".  
At first I tried returning `JSON.parse(JSON.stringify(myExportedFunction()))` - this worked, but only by converting the URL to a string. I ended up just changing my bookmarklet to construct a URL object itself from the URL string, and changed the "url" property to a string. So there are some limitations with `cloneInto` to be aware of.

**Comment by Aphexus** [comment]
So is this getting fixed? I have a simple bookmarklet that I can't use because of firefoxes stupid CSP. the bookmarklet does nothing but use fetch to a local sever to transfer the url. It was working with tampermonkey but I switched to VM because TM seems to have performance issues in some cases and now the bookmarklet doesn't work.

I can't add functions to window with unsafeWindow with VM which would allow me to move the bookmarklet code inside a script. I guess I'm stuck with TM ;/

**Comment by jaw-sh** [comment]
Using `@inject-into page` on Facebook yields this error. I need to monkeypatch window.WebSocket and this is making it impossible to do so. I assume they're deliberately breaking something. On FireFox.

**Comment by erusc** [comment]
I may have found another workaround.
Normally on https://developer.mozilla.org/en-US/, attempt to override `window.origin` would be blocked by CSP.
However I found that I can do this by injecting `script` element with data URI as `src` with [GM_addElement](https://violentmonkey.github.io/api/gm/#gm_addelement).
```js
// ==UserScript==
// @name    CSP Workaround
// @run-at  document-start
// @grant   GM_addElement
// @include *://developer.mozilla.org/*
// ==/UserScript==

// Blocked by CSP
Object.defineProperty(window,'origin',{value:'Hello Wrold'});

// Blocked by CSP
// <script>Object.defineProperty(window,'origin',{value:'Hello Wrold'})</script>
GM_addElement(document.documentElement,'script',{textContent:`Object.defineProperty(window,'origin',{value:'Hello Wrold'})`});

// Worked!!
// <script src="data:;charset=utf-8,Object.defineProperty(window,'origin',{value:'Hello Wrold'})"></script>
GM_addElement(document.documentElement,'script',{src:`data:;charset=utf-8,Object.defineProperty(window,'origin',{value:'Hello Wrold'})`});
```

The following script overrides XMLHttpRequest.prototype.open.
You need semicolons at line ends and have to excape some special characters like `#`.
An alternative is to encode the snippet to be inserted as base64 data URI with tools like https://www.toolsley.com/datauri.html.
And don't forget to add `@grant GM_addElement`.

```js
// ==UserScript==
// @name          Hook XHR#open
// @version       0.0.1
// @include       https://twitter.com/*
// @include       https://github.com/*
// @include       https://*.google.tld/*
// @grant         GM_addElement
// ==/UserScript==

GM_addElement('script',{src:`data:;charset=utf-8,

const xhrProto = window.XMLHttpRequest.prototype;

function hookXHROpen (oldOpen) {
    return function open () {
        console.warn('inside XHR%23open');
        return oldOpen.apply(this, arguments);
    }
}

xhrProto.open = hookXHROpen(xhrProto.open);

`})
```


It might still be blocked if the CSP header disallows any script though.
```
Content-Security-Policy: script-src 'none'
```

**Comment by hydrargyrum** [comment]
it seems greasemonkey is able to run scripts despite CSP? see https://github.com/greasemonkey/greasemonkey/issues/3228#issuecomment-4498409028
also, webextensions can modify CSP, so maybe it can be integrated in order to bypass it

**Comment by tophf** [comment]
Violentmonkey also runs scripts in CSP-restricted such pages and similarly to Greasemonkey runs them in an isolated world. In both cases the script has to use wrappedJSObject, exportFunction, and cloneInto as described in [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts).

Note that removing the CSP header reduces security of the webpage, so it's not something to be applied freely. Those who really want it can install a separate extension for that purpose or even disable CSP in Firefox globally via about:config.

**Comment by hydrargyrum** [comment]
> Violentmonkey also runs scripts in CSP-restricted such pages and similarly to Greasemonkey runs them in an isolated world. In both cases the script has to use wrappedJSObject, exportFunction, and cloneInto as described in [MDN](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts).

I guess this comes with limitations on what scripts can do, right? Are they documented? For some scripts (just adding a span somewhere in document) I was blocked by CSP and had to disable it temporarily.

**Comment by tophf** [comment]
I need to see the actual script and optionally run it on the page to say what happened, but generally DOM manipulations are not blocked.

**Comment by tophf** [comment]
Note though, if you use innerHTML or similar methods to add content, it may be indeed blocked by CSP trusted-types directives. The correct solution is to use direct DOM construction, e.g. using a simple lightweight DOM framework or implementing one yourself, or you can spoof/disable trustedTypes API of the page, but that's dirty and potentially unsafe.

**Comment by tophf** [comment]
The [latest release](https://github.com/violentmonkey/violentmonkey/releases/) adds `Bypass CSP in Firefox` option which modifies CSP of sites like github.com to run scripts in the `page` context so they can have direct access to JS environment without the need for `wrappedJSObject`.

**Comment by AlttiRi** [comment]
For me it does not work on [x.com](https://x.com/) with the simple `alert(1);` code.


```js
// ==UserScript==
// @name        alert page inject test
// @namespace   browser.scripts
// @match       https://x.com/*
// @match       https://github.com/*
// @match       https://web.telegram.org/*
// @version     0.0.1
// @description 2026-08-23
// @grant       none
// @inject-into page
// ==/UserScript==

alert(1);

```

VM 2.48.0 + Firefox 154 + "Bypass CSP in Firefox" → "Could not inject some scripts."

**Comment by tophf** [comment]
@AlttiRi, it works for me, so I can't investigate. Could be a conflict with another extension.

**Comment by AlttiRi** [comment]
I just tested a bit more.
The script above works on https://github.com/ **only** with `"Bypass CSP in Firefox"`.
On https://x.com/ it works without `"Bypass CSP in Firefox"` when you **not** logged in.
But, when you are logged in, the script does not work on https://x.com/ even with `"Bypass CSP in Firefox"` and all other extensions are disabled. There is no problem with Tampermonkey.


---

UPD:
For example, opening of a page:

- https://x.com/i/grok — `alert` works
- https://x.com/i/spaces/start — `alert` works
- https://x.com/home — does not work even with `bypass`
- https://x.com/spacex — does not work even with `bypass`

Also, https://web.telegram.org/a/ does not work (`/a/`, not `/k/` version).

**Comment by tophf** [comment]
Works for me both when signed-in and not...

**Comment by AlttiRi** [comment]
Maybe I'm missing something, but the most simple approach of injecting user scripts with `tabs.executeScript`\* ...just works, without any tricks.

```js
// background.js
browser.browserAction.onClicked.addListener((tab) => {
  browser.tabs.executeScript(tab.id, {file: "content.js"});
});

// content.js
void (async function injectUserCode() {
  const {userCode} = await browser.storage.local.get("userCode");

  const DEFAULT_CODE = `console.log("__userScriptBypassWorked"); window.__userScriptBypassWorked = true;`;
  const code = userCode || DEFAULT_CODE;

  const blob = new Blob([code], {type: "application/javascript"});
  const blobUrl = URL.createObjectURL(blob);

  const script = document.createElement("script");
  script.src = blobUrl;
  script.onload = function() {
    this.remove();
    URL.revokeObjectURL(blobUrl);
  };
  (document.head || document.documentElement).append(script);
})();
```

It works on pages with CSP, it works fine with a code is `fetch`ing data with service workers\*\*.

It just works the same way as this extension works in Chromium browsers.

---

\* ...which you used before 2019 (if I'm correct)
- https://github.com/violentmonkey/violentmonkey/issues/374

\** ...like this one https://greasyfork.org/en/scripts/446342, which does not work in FF with VM: 
- https://github.com/Neet-Nestor/Telegram-Media-Downloader/issues/94

but that userscript works with this simple inject approach in FF. I only added the extra line with `const unsafeWindow = window;` text.

This script https://github.com/QuantumDeus/x-media-grid-restore/blob/main/inject.js also is working in FireFox only with this inject approach. For this script instead of `browserAction.onClicked` I used:
```js
  "content_scripts": [
    {
      "matches": ["https://x.com/*"],
      "js": ["content.js"],
      "run_at": "document_start"
    }
  ]
```

**Comment by tophf** [comment]
A blob URL is executed asynchronously which makes document-start impossible.

**Comment by tophf** [comment]
Note that the current implementation doesn't support `<meta>` tags with CSP inside which is used by telegram website. I'll try copying my implementation from Stylus.

**Comment by AlttiRi** [comment]
Seems, patching the headers (`"content-security-policy"`, `"content-security-policy-report-only"`(?)) does not work on `x.com` because of the service worker.

When you open a page with the **registered** SW, there is no event with `"main_frame"` type in `webRequest.onHeadersReceived`, so, you can't patch the CSP header by adding the injected script SHA-256 _(if you listen only `"main_frame"` and `"sub_frame"`)_. _(...or read the correct `nonce`)_

Opening of a new _(handled with a SW)_ tab is **`"xmlhttprequest"`** type in Firefox, not `"main_frame"`.

With `["main_frame", "sub_frame" /*?*/, "xmlhttprequest"]` filter types in `onHeadersReceived` appending of `sha256-...` of the injected script works OK in `x.com`.

---

BTW, why not to add a meta parameter (`@inject-type`) _(for Firefox only?)_ to choose the inject approach?

The approach with blob-src script injecting is very simple, it bypasses both types of CSP (the header and `meta[http-equiv]`) and it's fine for 99+ % of userscripts.

**Comment by tophf** [comment]
Here x.com fails to register its service worker in Firefox for an unclear reason, which is why I couldn't replicate the problem apparently. Adding yet another injection mode with its own timing quirks doesn't seem good to me. I expect the proper solution would be to use the new browser.userScripts API which may be what Tampermonkey uses since it works as you say.

**Comment by tophf** [comment]
Actually we can't use blob URL approach for normal userscripts because it would require us to unsafely expose our GM functions and internal API to the webpage asynchronously that can be hijacked easily by any script. We have to do it synchronously in a specialized fashion that prevents triggering any events between the moment the API is exposed and the userscript in the webpage is executed. The only scripts that can utilize blob injection are those with `// @unwrap` and without document-start or document-body.

**Comment by kmantel** [comment]
> [@AlttiRi](https://github.com/AlttiRi), it works for me, so I can't investigate. Could be a conflict with another extension.

There's a conflict with the Stylus "Patch CSP to allow style assets" setting, testing on github.com with firefox developer edition. Disabling Stylus entirely doesn't fix it on my main profile though.
